trino/trino.go

```diff
@@ -1642,6 +1730,62 @@ func handleResponseError(status int, respErr ErrTrino) error {
 	}
 }
 
+func (qr *driverRows) startSegmentProcessor() {
+	go func() {
+		defer close(qr.stmt.spoolingRowsChannel)
+		defer close(qr.stmt.spoolingProcesserDone)
+		defer close(qr.stmt.downloadPermitCh)
+		buffer := make(map[int64][]queryData)
+		var nextExpectedOffset int64 = 0
+		maxBacklog := 2 * qr.stmt.spoolingMaxOutOfOrderSegments
+		backlogCount := 0
+
+		for {
+			var downloadPermitCh <-chan struct{}
+			if backlogCount <= maxBacklog {
```

Comment:
You should not need to check this explicitly, the size of the `downloadPermitCh` should control this - it will block downloaders when it gets full.